File: trunk/geopdes_base/inst/operators/op_gradu_gradv.m
% OP_GRADU_GRADV: assemble the stiffness matrix A = [a(i,j)], a(i,j) = (epsilon grad u_j, grad v_i).
%
%   mat = op_gradu_gradv (spu, spv, msh, epsilon);
%   [rows, cols, values] = op_gradu_gradv (spu, spv, msh, epsilon);
%
% INPUT:
%    
%   spu:     structure representing the space of trial functions (see sp_bspline_2d_phys)
%   spv:     structure representing the space of test functions  (see sp_bspline_2d_phys)
%   msh:     structure containing the domain partition and the quadrature rule (see msh_push_forward_2d)
%   epsilon: diffusion coefficient
%
% OUTPUT:
%
%   mat: assembled stiffness matrix
%   rows:   row indices of the nonzero entries
%   cols:   column indices of the nonzero entries
%   values: values of the nonzero entries
% 
% Copyright (C) 2009, 2010 Carlo de Falco
% Copyright (C) 2011, Rafael Vazquez
%
%    This program is free software: you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation, either version 3 of the License, or
%    (at your option) any later version.

%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program.  If not, see <http://www.gnu.org/licenses/>.

function varargout = op_gradu_gradv (spu, spv, msh, coeff, element_list)

  if (nargin == 4)
    element_list = 1:msh.nel;
  end
  nel = numel (element_list);

  gradu = reshape (spu.shape_function_gradients, spu.ncomp, [], ...
		   msh.nqn, spu.nsh_max, nel);
  gradv = reshape (spv.shape_function_gradients, spv.ncomp, [], ...
		   msh.nqn, spv.nsh_max, nel);

  ndir = size (gradu, 2);

  rows = zeros (nel * spu.nsh_max * spv.nsh_max, 1);
  cols = zeros (nel * spu.nsh_max * spv.nsh_max, 1);
  values = zeros (nel * spu.nsh_max * spv.nsh_max, 1);

  ncounter = 0;
  for iel = 1:nel
    iel_glob = element_list (iel);
    if (all (msh.jacdet(:, iel_glob)))
      jacdet_weights = msh.jacdet(:, iel_glob) .* ...
                       msh.quad_weights(:, iel_glob) .* coeff(:, iel_glob);
      jacdet_weights = repmat (jacdet_weights, [1,spu.nsh(iel)]);

      gradu_iel = permute (gradu(:, :, :, 1:spu.nsh(iel), iel), [1 2 4 3]);
      gradu_iel = reshape (gradu_iel, spu.ncomp * ndir, spu.nsh(iel), msh.nqn);
      gradu_iel = permute (gradu_iel, [1 3 2]);

      gradv_iel = permute (gradv(:, :, :, 1:spv.nsh(iel), iel), [1 2 4 3]);
      gradv_iel = reshape (gradv_iel, spv.ncomp * ndir, spv.nsh(iel), msh.nqn);
      gradv_iel = permute (gradv_iel, [1 3 2]);

      for idof = 1:spv.nsh(iel)
        ishg = repmat (gradv_iel(:, :, idof), [1 1 spu.nsh(iel)]);

        rows(ncounter+(1:spu.nsh(iel))) = spv.connectivity(idof, iel);
        cols(ncounter+(1:spu.nsh(iel))) = spu.connectivity(1:spu.nsh(iel), iel);

        values(ncounter+(1:spu.nsh(iel))) = sum (jacdet_weights .* ...
          reshape (sum (ishg .* gradu_iel, 1), msh.nqn, spu.nsh(iel)), 1);
        ncounter = ncounter + spu.nsh(iel);
      end
    else
      warning ('geopdes:jacdet_zero_at_quad_node', 'op_gradu_gradv: singular map in element number %d', iel)
    end
  end

  if (nargout == 1)
    varargout{1} = sparse (rows, cols, values, spv.ndof, spu.ndof);
  elseif (nargout == 3)
    varargout{1} = rows;
    varargout{2} = cols;
    varargout{3} = values;
  else
    error ('op_gradu_gradv: wrong number of output arguments')
  end

end

File: branches/parallel_support/geopdes_base/src/op_gradu_gradv.cc
/* Copyright (C) 2009 Carlo de Falco

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.
 
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#include "geopdes.h"

DEFUN_DLD(op_gradu_gradv, args, nargout,"\n\
OP_GRADU_GRADV: assemble the stiffness matrix A = [a(i,j)], a(i,j) = (epsilon grad u_j, grad v_i).\n\
\n\
  mat = op_gradu_gradv (spu, spv, msh, epsilon);\n\
\n\
INPUT: \n\
\n\
  spu:     structure representing the space of trial functions (see sp_bspline_2d_phys) \n\
  spv:     structure representing the space of test functions  (see sp_bspline_2d_phys) \n\
  msh:     structure containing the domain partition and the quadrature rule (see msh_push_forward_2d) \n\
  epsilon: diffusion coefficient \n\
\n\
OUTPUT: \n\
\n\
  mat: assembled stiffness matrix \n\
")
{
  
  octave_value_list retval;

  geopdes_mesh  msh (args(2).map_value ());
  geopdes_space spu (args(0).map_value (), msh);
  geopdes_space spv (args(1).map_value (), msh);
  Matrix coeff = args(3).matrix_value();

  if (!error_state)
    {
      
      dim_vector dims (msh.nel () * spv.nsh_max () * spu.nsh_max (), 1);
      Array <octave_idx_type> I (dims, 0);
      Array <octave_idx_type> J (dims, 0);
      Array <double> V (dims, 0.0);
     
      SparseMatrix mat;
        
#pragma omp parallel default (none) shared (msh, spu, spv, I, J, V, coeff)
      {
        octave_idx_type counter;
#pragma omp for
        for ( octave_idx_type iel=0; iel < msh.nel (); iel++) 
          if (msh.area (iel) > 0.0)
            {
              for ( octave_idx_type idof(0); idof < spv.nsh (iel); idof++) 
                {
                  for ( octave_idx_type jdof(0); jdof < spu.nsh (iel); jdof++) 
                    {

                      counter = jdof + spu.nsh (iel) * (idof + spv.nsh (iel) * iel);

                      I(counter) = spv.connectivity (idof, iel) - 1;
                      J(counter) = spu.connectivity (jdof, iel) - 1;
                      V(counter) = 0.0;
                      for ( octave_idx_type inode(0); inode < msh.nqn (); inode++)
                        {
                          if (msh.weights (inode, iel) > 0.0)
                            {
                              double s = 0.0;
                              for (octave_idx_type icmp(0); icmp< spu.ncomp (); icmp++)
                                for (octave_idx_type idir(0); idir < msh.ndir (); idir++)
                                  s += spv.shape_function_gradients (icmp, idir, inode, idof, iel) * 
                                    spu.shape_function_gradients (icmp, idir, inode, jdof, iel);
                          
                              V(counter) += 
                                msh.jacdet (inode, iel) *
                                msh.weights (inode, iel) *
                                coeff(inode, iel) *
                                s;			  
                            }  
                        } // end for inode
                      //		    if (idof != jdof) // copy upper triangular part to lower
                      //		      { 
                      //		        I(counter) = J(counter-1);
                      //		        J(counter) = I(counter-1);
                      //		        V(counter) = V(counter-1);
                      //		        counter++;
                      //		      } 
                    } // end for jdof
                } // end for idof
            } else {
#pragma omp critical
            {warning_with_id ("geopdes:zero_measure_element", "op_gradu_gradv: element %d has 0 area (or volume)", iel);}
          }  // end for iel, if area > 0
      } // end of openmp parallel section

      mat = SparseMatrix (V, I, J, spv.ndof (), spu.ndof (), true);
      retval(0) = octave_value (mat);

    } // end if !error_state
  return retval;
}

